File: python/src/imc.py
"""Top-level main program for fcimc."""

import argparse
import logging
import random
import time as pytime

import imc_global_part_data as part

import imc_mesh
import imc_opcon
import imc_user_input


def setup_logger():
    """Set up logging."""
    logger = logging.getLogger(__name__)
    logger.setLevel(logging.DEBUG)
    console_handler = logging.StreamHandler()
    console_handler.setLevel(logging.DEBUG)
    formatter = logging.Formatter("%(asctime)s %(name)24s %(levelname)8s: %(message)s")
    console_handler.setFormatter(formatter)
    logger.addHandler(console_handler)
    return logger


def parse_args():
    """Parse command-line arguments and options."""
    parser = argparse.ArgumentParser(
        description="Python implementation of Fleck and Cummings (1971) implicit Monte Carlo code."
    )

    parser.add_argument("-i", "--input", default="fcimc.in", help="Name of input file")
    parser.add_argument(
        "-o", "--output", default="fcimc.out", help="Name of output file"
    )
    parser.add_argument("-d", "--debug", default=False, help="Debug mode")

    return parser.parse_args()


def main(input_file, output_file, debug_mode):
    """
    @brief   Top-level function for fcimc.

    @details Can be called within Python after importing, so has simple/flat signature.

    @param   input_file
    @param   output_file
    @param   debug_mode
    """
    tm0 = pytime.perf_counter()

    logger = setup_logger()

    rng_seed = 12345
    # logger.info("Initialising RNG with seed %d", rng_seed)
    random.seed(rng_seed)

    part.n_max = 20000

    imc_user_input.read(input_file)
    imc_user_input.echo()

    imc_mesh.make()
    # imc_mesh.echo()

    imc_opcon.run(output_file)

    tm1 = pytime.perf_counter()
    print("Time taken for calculation = {:10.2f} s".format(tm1 - tm0))


if __name__ == "__main__":

    # Command-line options
    args = parse_args()

    # Call the main function
    main(args.input, args.output, args.debug)

File: fortran/src/imc.f90
subroutine usage(routine_name)

    implicit none

    character(len=*), intent(in) :: routine_name

    write(*,'(3A)') 'usage: python ', routine_name, &
                    ' [-d | --debug]'// &
                    ' [-h | --help]'// &
                    ' [(-i | --input) INPUT_FILE_NAME]'// &
                    ' [(-o | --output) OUTPUT_FILE_NAME]'

end subroutine usage


subroutine parse_args()

    use imc_global_io_data,   only: input_file, output_file

    implicit none

    character(len=256) :: routine_name
    logical            :: print_help
    logical            :: debug_mode

    character(len=256) :: cmdline_arg
    integer(kind=4)    :: cmdline_len
    integer(kind=4)    :: cmdline_stat
    integer(kind=4)    :: iarg

    input_file = 'fcimc.in'
    output_file = 'fcimc.out'
    print_help = .false.
    debug_mode = .false.

    ! Fortran 2003 standard (get_command_argument)...
    iarg = 1
    do
        call get_command_argument(iarg, cmdline_arg, cmdline_len, cmdline_stat)
        if (cmdline_stat > 0) exit

        select case (trim(cmdline_arg))
        case ('-i', '--input')
            iarg = iarg + 1
            call get_command_argument(iarg, cmdline_arg, cmdline_len, cmdline_stat)
            if (cmdline_stat /= 0) then
                write(*,*) 'Error reading input file name'
                stop
            endif
            input_file = trim(cmdline_arg)
        case ('-o', '--output')
            iarg = iarg + 1
            call get_command_argument(iarg, cmdline_arg, cmdline_len, cmdline_stat)
            if (cmdline_stat /= 0) then
                write(*,*) 'Error reading output file name'
                stop
            endif
            output_file = trim(cmdline_arg)
        case ('-h', '--help')
            print_help = .true.
        case ('-d', '--debug')
            debug_mode = .true.
        case default
            write(*,*) 'Unrecognised command line option: ', trim(cmdline_arg)
            stop
        end select
        iarg = iarg + 1
    enddo

    if (print_help) then
      call get_command_argument(0, routine_name)
      call usage(trim(routine_name))
      stop
    endif

end subroutine parse_args


program main
    ! Top-level main program for fcimc.
    use imc_global_part_data, only: n_max, xpos, icell, nrg, startnrg, dead, &
                                    t, mu, frq, origin, inext
    use imc_utils,            only: init_random_seed

    implicit none

    integer       :: tcom, tfin, clock_rate
    integer       :: rng_seed

    call system_clock(tcom)

    call parse_args()

    rng_seed = 12345
    call init_random_seed(rng_seed) ! Non-zero argument fixes random number sequence - reproducible

    n_max = 20000

    call imc_user_input_read
    call imc_user_input_echo

    allocate(origin(n_max))
    allocate(xpos(n_max))
    allocate(t(n_max))
    allocate(mu(n_max))
    allocate(nrg(n_max))
    allocate(startnrg(n_max))
    allocate(frq(n_max))
    allocate(icell(n_max))
    allocate(dead(n_max))
    origin(:)   = 0
    xpos(:)     = 0.d0
    t(:)        = 0.d0
    mu(:)       = 0.d0
    nrg(:)      = 0.d0
    startnrg(:) = 0.d0
    frq(:)      = 0.d0
    icell(:)    = 0
    dead(:)     = .true.
    inext       = 1

    call imc_mesh_make

    call imc_opcon

    deallocate(origin)
    deallocate(xpos)
    deallocate(t)
    deallocate(mu)
    deallocate(nrg)
    deallocate(startnrg)
    deallocate(frq)
    deallocate(icell)
    deallocate(dead)

    call system_clock(tfin, clock_rate)
    write(*,'(A,F10.2,A)') 'Time taken for simulation = ', &
        & float(tfin - tcom) / float(clock_rate), ' s'

end program main
